### PR TITLE
Add Azure storage native support

### LINKED ISSUE #29: Add Azure storage native support
Hi,

There is no native support for Azure storage, which makes it difficult to store backups in Azure in a Juju-managed way. Currently available options:
1. Duplicity provides AWS S3 integration, but the Azure storage account doesn't have native AWS S3 interface support. 
2. Also, there is an option to use the SFTP endpoint, but it increases the cost of Azure storage for cold backups from some 10-15 USD per month to 200+ USD per month as it requires a VM provisioned.

Azure solutions lack a Juju-managed way to manage backups. I think Charmed Duplicity may be a way forward.

### PR DESCRIPTION
**Pull request by ashugauttam225**
i have added the functionality for native support for Azure storage, which makes it available to store backups in Azure in a Juju-managed way. 
Please review and do comment and i will make the necessary changes accordingly. 
I will add the unit test later. 

Fixes: #29

### CODE REVIEW

File: src/config.yaml [hunk 1]
```diff
@@ -42,6 +46,12 @@ options:
       Secret access key for the AWS IMA user. The user must have a policy that
       grants it privileges to upload to the S3 bucket. This value is required
       when backend='s3'.
+  azure_connection_string:
+    type: string
+    default: ""
+    description: |
```
Comment (jneo8) [inline src/config.yaml]: **suggestion (non-blocking):** 

Can you provide some example string here or a link to document where people can find the example.

File: src/lib/lib_duplicity.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def _backup_url(self):
                 else:
                     url += "{}@".format(user)
             url += remote_path.replace(prefix, "")
-        elif backend in ["s3", "file"]:
+        elif backend in ["s3", "file","azure"]:
```
Comment (jneo8) [inline src/lib/lib_duplicity.py]: **nitpick (blocking):** 

Need a space here

```python
elif backend in ["s3", "file", "azure"]
```

File: src/reactive/duplicity.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def install_duplicity():
     fetch.apt_install("python-paramiko")
     fetch.apt_install("python-boto")
     fetch.apt_install("lftp")
+    os.system("pip install azure-storage-blob")
```
Comment (jneo8) [inline src/reactive/duplicity.py]: **nitpick (blocking):** 

Is this always required? Also `os.system` is not a recommended way to run pip install. I suggest move dependency to requirements.txt so the package will be included when charm is building.

**Review by jneo8 — CHANGES_REQUESTED** [review]
Thanks for the contribution.
Please include unit tests in the PR, thanks.

File: src/lib/lib_duplicity.py (hunk 2 above)
Comment (sudeephb) [reply src/lib/lib_duplicity.py]: +1
To avoid issues like these, you can run `make black` and `make lint`(and fix if it complains), before pushing the changes.

File: src/config.yaml (hunk 1 above)
Comment (ashugauttam225) [reply src/config.yaml]: sure, will add it.

File: src/lib/lib_duplicity.py (hunk 2 above)
Comment (ashugauttam225) [reply src/lib/lib_duplicity.py]: thanks, will do it from now on.
TIL

File: src/reactive/duplicity.py (hunk 3 above)
Comment (ashugauttam225) [reply src/reactive/duplicity.py]: yes, It is required, actually i tried adding it in requirements.txt at the first place, but it's not working as expected. i am doing local charmcraft pack and then deploy. 
It gives "No module named 'azure.storage.blob" error.
Any advice ?

Comment (jneo8) [reply src/reactive/duplicity.py]: After checking the behavior of charm.

The charm use subprocess to call `duplicity`, which is installed by apt. So the `azure-storage-blob` has to be install in system level python.

Sorry for the misleading.

But here we should not use `os.system`, we can use `subprocess` to replace it and also I will make it as a separate function like `def install_in_system_python` for better testable in the unittest.

File: src/templates/periodic_backup
Comment (dparv) [inline src/templates/periodic_backup:1]: juju 2 vs juju 3 compatibility?

File: src/config.yaml (hunk 1 above)
Comment (ashugauttam225) [reply src/config.yaml]: added it.

File: src/lib/lib_duplicity.py (hunk 2 above)
Comment (ashugauttam225) [reply src/lib/lib_duplicity.py]: done

File: src/reactive/duplicity.py (hunk 3 above)
Comment (ashugauttam225) [reply src/reactive/duplicity.py]: added as adviced.

File: src/templates/periodic_backup
Comment (ashugauttam225) [reply src/templates/periodic_backup:1]: added something as a novice. please review and let me know if it can be trusted or i can also check "/var/lib/juju/tools/ " to fetch the juju version. and then update the template.

File: src/lib/lib_duplicity.py [hunk 4]
```diff
@@ -185,10 +197,16 @@ def _render_deletion_cron(self):
             deletion_frequency = "40 * * * *"
         elif deletion_frequency == "daily":
             deletion_frequency = "0 23 * * *"
+        is_juju3 = subprocess.getoutput("ls /usr/bin/ | grep juju-exec")
+        if is_juju3:
+            juju_binary = "juju-exec"
+        else:
+            juju_binary = "juju-run"
```
Comment (jneo8) [inline src/lib/lib_duplicity.py]: **nitpick (non-blocking):** 

`juju_binary` is not precise.

Maybe `juju_exec_binary`.

File: src/lib/lib_duplicity.py [hunk 5]
```diff
@@ -166,10 +172,16 @@ def _render_backup_cron(self):
...
         cron_ctx = {
             "frequency": backup_frequency,
             "unit_name": hookenv.local_unit(),
             "charm_dir": hookenv.charm_dir(),
+            "juju_command": juju_binary,
```
Comment (jneo8) [inline src/lib/lib_duplicity.py]: **nitpick:** 

Why don't use the same name as `juju_exec_binary`

File: src/reactive/duplicity.py [hunk 6]
```diff
@@ -40,6 +41,18 @@
...
+def install_in_system_python(package_to_install):
+    """Install dependency in system python.
+
+    The charm use subprocess to call duplicity,
+    which is installed by apt.
+    So the azure-storage-blob has to be installed n system level python
+    """
+    subprocess.run(["sudo", "pip", "install", package_to_install])
+
+    set_flag("duplicity.azure_package_installed")
```
Comment (jneo8) [inline src/reactive/duplicity.py]: **nitpick (blocking):** 

Not a nice place to run `set_flag`, this function should be more low level and only related to run subprocess.

Also error handling is required.

File: src/reactive/duplicity.py [hunk 7]
```diff
@@ -56,6 +69,7 @@ def install_duplicity():
     fetch.apt_install("python-paramiko")
     fetch.apt_install("python-boto")
     fetch.apt_install("lftp")
+    install_in_system_python("azure-storage-blob")
```
Comment (jneo8) [inline src/reactive/duplicity.py]: **nitpick:** 

Is the package has to be installed even it's not azure?
Maybe you want to move this step to `config-changed` event handler?

Also you should consider the issue if network is not available.


```python
ok = install_in_system_python("azure-storage-blob")
if ok:
    set_flag("duplicity.azure_package_installed")
else:
    # raise error so user can retry with `juju-resolve`
```

**Review by jneo8 — CHANGES_REQUESTED** [review]
Also please include unit tests in the PR.

File: src/lib/lib_duplicity.py (hunk 5 above)
Comment (chanchiwai-ray) [reply src/lib/lib_duplicity.py]: **nitpick**: maybe consider making it a short private function, and add unit test for that.

e.g.
```python
def _get_juju_exec_binary(self):
    ...
```

File: src/reactive/duplicity.py (hunk 6 above)
Comment (chanchiwai-ray) [reply src/reactive/duplicity.py]: ^👍🏼 to James

this function name suggests nothing about `duplicity.azure_package_installed`, this line should be somewhere else

File: src/reactive/duplicity.py [hunk 8]
```diff
@@ -102,6 +124,12 @@ def validate_backend():
             return
 
 
+@when_not("duplicity.azure_package_installed")
+def install_azure_package():
+    """If charm is refreshed it will install the azure package."""
+    install_in_system_python("azure-storage-blob")
+
```
Comment (chanchiwai-ray) [inline src/reactive/duplicity.py]: The `set_flag("duplicity.azure_package_installed")` should be here. But to avoid complication, I think you can remove this logic altogether, and simply add a single call `packages.pip_install("azure-storage-blob")` to `install_duplicity()` and update the docstring to include azure as a dependency

File: src/reactive/duplicity.py (hunk 3 above)
Comment (chanchiwai-ray) [reply src/reactive/duplicity.py]: Can you use 
```python
from charmhelpers.fetch.python import packages
...
packages.pip_install("...")
```
instead

File: src/config.yaml [hunk 9]
```diff
@@ -9,14 +9,17 @@ options:
...
   remote_backup_url:
     type: string
     default: ""
     description: |
       URL to the remote server and its local path to be used as the
-      backup destination.
+      backup destination. 
+      While using azure as a backend:
+      Duplicity will take care to create the container 
+      when performing the backup.  Do not create it manually before.
```
Comment (samuelallan72) [inline src/config.yaml:22]: What happens if the container is already created?

File: src/lib/lib_duplicity.py (hunk 5 above)
Comment (samuelallan72) [inline src/lib/lib_duplicity.py]: This appears to be unrelated to azure support, as it works on adding juju 3.x support.  Could you move these changes to a separate PR please?

**Review by samuelallan72 — CHANGES_REQUESTED** [review]

File: src/lib/lib_duplicity.py (hunk 5 above)
Comment (ashugauttam225) [reply src/lib/lib_duplicity.py]: sure will update.

File: src/reactive/duplicity.py (hunk 3 above)
Comment (ashugauttam225) [reply src/reactive/duplicity.py]: It is not working as expected, so for now i have to use subprocess.

File: src/lib/lib_duplicity.py (hunk 4 above)
Comment (ashugauttam225) [reply src/lib/lib_duplicity.py]: i will add this functionality in another PR

File: src/lib/lib_duplicity.py (hunk 5 above)
Comment (ashugauttam225) [reply src/lib/lib_duplicity.py]: sure will do it in another PR.

File: src/reactive/duplicity.py (hunk 6 above)
Comment (ashugauttam225) [reply src/reactive/duplicity.py]: changed this logic.

File: src/config.yaml (hunk 9 above)
Comment (ashugauttam225) [reply src/config.yaml:22]: not a problem, it will start putting data inside it, but if you explicitly set the config with some name for example: remote_backup_url = azure:///some-dir

then it would throw an error, "invalid container name" 
because, duplicity adds the  container-name as current unit name. so now the backed url will become - azure:///some-dir/duplicity-<id> and the error would be : "Invalid Azure Storage Blob container name"

so if container is already created on azure then it would be fine. but we should not explicitly set the name in the config.

File: src/lib/lib_duplicity.py (hunk 5 above)
Comment (samuelallan72) [reply src/lib/lib_duplicity.py]: Thanks :)

File: src/config.yaml (hunk 9 above)
Comment (samuelallan72) [reply src/config.yaml:22]: Thanks for the explanation - could you add this explanation to the docs here?

File: src/reactive/duplicity.py (hunk 7 above)
Comment (jneo8) [reply src/reactive/duplicity.py]: hi @ashugauttam225 , the error handling is still missing here.
Do you want to have that in this PR or we just raise error and let charm get into error status?

**Review by jneo8 — CHANGES_REQUESTED** [review]
Just leave a message to confirm the scope of this PR.
And do we want to make juju2/3 compatible in this PR? Or you will provide another PR to fix it?

**Comment by jneo8** [comment]
Hi @ashugauttam225 , are you still work on this one?

**Review by jneo8 — DISMISSED** [review]

**Review by chanchiwai-ray — DISMISSED** [review]

**Review by samuelallan72 — APPROVED** [review]

**Review by chanchiwai-ray — APPROVED** [review]
